main/src/ca/uwaterloo/flix/language/phase/MonoDefs.scala

```diff
@@ -120,46 +121,44 @@ object MonoDefs {
       * - No type aliases
       * - Types are normalized (e.g. record ordering)
       */
-    def apply(tpe0: Type): Type = {
-      // NB: The order of cases has been determined by code coverage analysis.
-      def visit(t: Type): Type =
-        t match {
-          // When a substitution is performed, eliminate variables from the result.
-          case x: Type.Var => s.m.get(x.sym) match {
-            case Some(tpe) => tpe.map(default)
-            case None => default(t)
-          }
-          case Type.Cst(_, _) => t
-          case Type.Apply(t1, t2, loc) =>
-            val y = visit(t2)
-            visit(t1) match {
-              // Simplify boolean equations.
-              case Type.Cst(TypeConstructor.Complement, _) => Type.mkComplement(y, loc)
-              case Type.Apply(Type.Cst(TypeConstructor.Union, _), x, _) => Type.mkUnion(x, y, loc)
-              case Type.Apply(Type.Cst(TypeConstructor.Intersection, _), x, _) => Type.mkIntersection(x, y, loc)
-
-              case Type.Cst(TypeConstructor.CaseComplement(sym), _) => Type.mkCaseComplement(y, sym, loc)
-              case Type.Apply(Type.Cst(TypeConstructor.CaseIntersection(sym), _), x, _) => Type.mkCaseIntersection(x, y, sym, loc)
-              case Type.Apply(Type.Cst(TypeConstructor.CaseUnion(sym), _), x, _) => Type.mkCaseUnion(x, y, sym, loc)
-
-              // Else just apply
-              case x => Type.Apply(x, y, loc)
-            }
-          case Type.Alias(sym, args0, tpe0, loc) =>
-            val args = args0.map(visit)
-            val tpe = visit(tpe0)
-            Type.Alias(sym, args, tpe, loc)
-
-          // Perform reduction on associated types.
-          case Type.AssocType(cst, arg0, _, loc) =>
-            val arg = visit(arg0)
-            EqualityEnvironment.reduceAssocType(cst, arg, eqEnv) match {
-              case Ok(t) => t
-              case Err(_) => throw InternalCompilerException("unexpected associated type reduction failure", loc)
-            }
+    def apply(tpe0: Type): Type = tpe0 match {
+      case x: Type.Var =>
+        // Remove variables by substitution, otherwise by default type.
+        s.m.get(x.sym) match {
+          case Some(t) =>
+            // The image of s might have unconstrained variables and be non-normalized.
+            // It is important to use default first, because the variables of t might be in the
+            // domain of the substitution.
+            apply(t.map(default))
```

Comment:
First important change, apply must be used for normalization (doesn't do anything right now)